elastic/beats
# Add awscloudwatch filebeat input

## What does this PR do?

This PR is to add awscloudwatch input into Filebeat. [FilterLogEvents](https://docs.aws.amazon.com/AmazonCloudWatchLogs/latest/APIReference/API_FilterLogEvents.html) AWS API is used to get all log events from a given log group. 

The config for using `awscloudwatch` input looks like below:
```
filebeat.inputs:
  - type: awscloudwatch
    credential_profile_name: elastic-beats
    log_group_arn: arn:aws:logs:us-east-1:428152502467:log-group:test:*
    region: us-east-1
    scan_frequency: 30s
    start_position: beginning
    api_timeout: 5m
```

With this config, Filebeat will enable `awscloudwatch` input to collect all logs from log group `test` in region `us-east-1` (this info is parsed from the log group ARN) starting from the beginning of the log group and then check for new log events every 1-minute(defined by the `scan_frequency`).  

If users only wants to collect new log events/messages from now going forward, then `start_position` can be specified to be `end`. 

User can also specify a list of log streams under the log group to collect logs from or a `log_stream_prefix` to collect events only from log streams that have names starting with this prefix.

Sample output document:
```
{
  "_index": "filebeat-8.0.0-2020.06.26-000001",
  "_type": "_doc",
  "_id": "35535245334778869563151408522177877742806142229743009792",
  "_version": 1,
  "_score": null,
  "_source": {
    "@timestamp": "2020-06-29T18:31:50.000Z",
    "input": {
      "type": "awscloudwatch"
    },
    "ecs": {
      "version": "1.5.0"
    },
    "event.ingested": "2020-06-29T18:33:21.207Z",
    "message": "Apr 23 17:40:01 ip-172-31-81-156 systemd: Removed slice User Slice of root.",
    "log.file.path": "test/test1",
    "agent": {
      "version": "8.0.0",
      "ephemeral_id": "8b7ddd05-a81f-44c3-a0b2-c9a78396b97e",
      "id": "7578d49c-6588-4843-85cc-ad3859f99ed1",
      "name": "KaiyanMacBookPro",
      "type": "filebeat"
    },
    "awscloudwatch": {
      "event_id": "35535245334778869563151408522177877742806142229743009792",
      "log_group": "test",
      "log_stream": "test1",
      "ingestion_time": "2020-06-29T18:31:51.000Z",
      "timestamp": "2020-06-29T18:31:50.000Z"
    },
    "cloud": {
      "provider": "aws",
      "region": "us-east-1"
    }
}
```
## Why is it important?

This input allows user to collect logs from CloudWatch without sending them into S3 bucket with SQS setup for notification.

## Checklist




## How to test this PR locally
1. Go to AWS CloudWatch portal and click Logs https://console.aws.amazon.com/cloudwatch/home?region=us-east-1#logsV2:log-groups
2. Create a new log group and log stream
3. Click Action and Create log event
4. Start Filebeat with filebeat.yml contains awscloudwatch input like below:
```
filebeat.inputs:
  - type: awscloudwatch
    credential_profile_name: elastic-beats
    log_group_arn: arn:aws:logs:us-east-1:428152502467:log-group:test:*
    region: us-east-1
    scan_frequency: 30s
    start_position: beginning
    api_timeout: 5m
```
5. You should see logs in Kibana and if you create new log events in AWS CloudWatch, these new logs should be collected after the scan_frequency as well.

## Related issues

closes https://github.com/elastic/beats/issues/17292

## Screenshots
When setting start_position to beginning, all existing logs will be collected and then Filebeat will scan every 30 seconds(based on the `scan_frequency`):
<img width="1671" alt="Screen Shot 2020-06-23 at 4 56 48 PM" src="https://user-images.githubusercontent.com/14081635/85484866-e75a8080-b584-11ea-8586-b295260e9c09.png">
